### PR TITLE
Add RunfilesLibraryInfo

### PR DESCRIPTION
**Pull request by fmeum**
The new provider marks runfiles libraries as such and will be used by
both Bazel itself and language rules to emit additional information
required for executable targets to find their runfiles in the presence
of repository mappings.

Work towards #16124

### CODE REVIEW

**Comment by fmeum** [comment]
@Wyverald

**Comment by oquenchil** [comment]
Hey Fabian,

@Wyverald showed me the error you were getting with this in testNoRunfilesLibraryUsers. But I'm confused why you need this provider, I thought in the proposal we had concluded that we can use the existing runfiles provider. Why do you need to know if the dependency is actually giving you runfiles or not? If you do need to know, I think you can get to the runfiles from the DefaultInfo provider and check if the depset of files is empty or not.

**Comment by fmeum** [comment]
@oquenchil This is a bit confusing, so let me try to clarify the situation: The `RunfilesLibraryInfo` provider is meant to mark a dependency as providing a runfiles **library**, not merely (and not necessarily) runfiles. For example, once the proposal has been implemented, `@bazel_tools//tools/cpp/runfiles` will advertise this provider. Direct dependents of targets advertising this provider will be tracked as "runfiles library users" in the existing `RunfilesProvider` - this is what I initially thought I needed a second new provider (which I called `RunfilesLibraryUsersProvider`) for.

The `RunfilesLibraryInfo` is needed because e.g. Java rules need to emit additional actions to make the current repository name available to code as a constant - this should only happen for targets that actually need it. Furthermore, the list of runfiles library users can be used to trim down the repository mapping manifest, which allows for more efficient caching in the face of changes to third-party dependencies' repository mappings.

Thanks for looking into the test failure, please let me know if you want additional clarifications.

**Comment by oquenchil** [comment]
Thanks for the clarification Fabian. You are getting a Skyframe error and I'm not very familiar with it. @Wyverald will be asking around though. In the mean time, it might help them if you can upload the whole change that produces this error.

**Comment by fmeum** [comment]
@oquenchil I pushed the failing test as a new commit https://github.com/bazelbuild/bazel/pull/16126/commits/7304185e722d08a922e3abe493099f85ae35ac4f on #16126.

File: src/main/java/com/google/devtools/build/lib/analysis/BUILD [hunk 1]
```diff
@@ -340,6 +340,7 @@ java_library(
         ":rule_configured_object_value",
         ":rule_definition_environment",
         ":run_environment_info",
+        "runfiles_library_info",
```
[the remark this answers is not in the document]
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/analysis/BUILD]: Done

**Review by oquenchil — APPROVED** [review]
LGTM

**Comment by Wyverald** [comment]
@sgowroji could you import this? Thanks!

**Pull request closed without merging**